### PR TITLE
`ContainItemsAssignableTo` Now Expects At Least One Item Assignable to `T`

### PR DESCRIPTION
**Pull request by MullerWasHere**
Closes Issue https://github.com/fluentassertions/fluentassertions/issues/1618
Previously, `ContainItemsAssignableTo` behaved exactly the same as `AllBeAssignableTo`. `ContainItemsAssignableTo` has been changed to expect at least one item assignable to `T`.

## IMPORTANT 

* [X] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [X] The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
* [x] If the contribution adds a feature or fixes a bug, please update [**the release notes**](https://github.com/fluentassertions/fluentassertions/tree/master/docs), which are published on the [website](https://fluentassertions.com/releases).
* [ ] If the contribution changes the public API the changes needs to be included by running `AcceptApiChanges.ps1`/`AcceptApiChanges.sh`.
* [ ] If the contribution affects [the documentation](https://github.com/fluentassertions/fluentassertions/tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).

### CODE REVIEW

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 1]
```diff
@@ -965,20 +965,18 @@ public AndConstraint<TAssertions> ContainItemsAssignableTo<TExpectation>(string
 
             if (success)
             {
```
Comment (jnyrup) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:967]: `only` needs to be removed from the `<summary>`

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 2]
```diff
@@ -965,20 +965,18 @@ public AndConstraint<TAssertions> ContainItemsAssignableTo<TExpectation>(string
...
-                    ++index;
-                }
+                Execute.Assertion
+                    .BecauseOf(because, becauseArgs)
+                    .WithExpectation("Expected {context:collection} to contain element assignable to type {0}{reason}, ", typeof(TExpectation).FullName)
+                    .Given(() => Subject)
+                    .ForCondition(subject => subject.Any())
+                    .FailWith("but was empty.")
+                    .Then
+                    .ForCondition(subject => subject.Any(x => typeof(TExpectation).IsAssignableFrom(GetType(x))))
```
[the remark this answers is not in the document]
Comment (jnyrup) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:975]: FYI, this implementation is almost identical to the existing `AllBeAssignableTo`.

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.ContainItemsAssignableTo.cs [hunk 3]
```diff
@@ -24,7 +24,7 @@ public void Should_succeed_when_asserting_collection_with_all_items_of_same_type
         }
 
         [Fact]
-        public void Should_fail_when_asserting_collection_with_items_of_different_types_only_contains_item_of_one_type()
+        public void Should_succeed_when_asserting_collection_with_items_of_different_types_only_contains_item_of_one_type()
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.ContainItemsAssignableTo.cs:27]:
```suggestion
        public void Should_succeed_when_asserting_collection_with_items_of_different_types_contains_item_of_expected_type()
```

**Review by jnyrup — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MullerWasHere) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: @dennisdoomen the last changes in f4f87bad178ba8552c47743844ec0b35e61807af no longer enumerate the collection twice:
https://github.com/MullerWasHere/fluentassertions/blob/f4f87bad178ba8552c47743844ec0b35e61807af/Src/FluentAssertions/Collections/GenericCollectionAssertions.cs#L968-L976

@jnyrup Should I refactor both `AllBeAssignableTo` and `ContainItemsAssignableTo` to reduce duplication?

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 4]
```diff
@@ -965,20 +965,18 @@ public AndConstraint<TAssertions> ContainItemsAssignableTo<TExpectation>(string
...
+                    .FailWith("but found {0}.",
+                        subject => $"[{string.Join(", ", subject.Select(x => GetType(x).FullName))}]")
```
[the remark this answers is not in the document]
Comment (MullerWasHere) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: @dennisdoomen Done in https://github.com/fluentassertions/fluentassertions/commit/f4f87bad178ba8552c47743844ec0b35e61807af. Let me know if I should've done it any other way.

File: docs/_pages/releases.md [hunk 5]
```diff
@@ -12,6 +12,7 @@ sidebar:
 ### What's New
 
 ### Fixes
+* `ContainItemsAssignableTo` Now Expects At Least One Item Assignable to `T` - [#1765](https://github.com/fluentassertions/fluentassertions/pull/1765)
```
[the remark this answers is not in the document]
Comment (MullerWasHere) [reply docs/_pages/releases.md]: 🙃 Oops, changed in https://github.com/fluentassertions/fluentassertions/commit/f4f87bad178ba8552c47743844ec0b35e61807af 😅

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: > Should I refactor both AllBeAssignableTo and ContainItemsAssignableTo to reduce duplication?

No, they solve different problems.

[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: > the last changes in f4f87ba no longer enumerate the collection twice:

It still does. You first do an `Any()` and then an `Any(predicate)`

[the remark this answers is not in the document]
Comment (MullerWasHere) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: @dennisdoomen I misunderstood that as a potential issue with `IEnumerable`s. Removed `any()` and adjusted unit tests in 60a889020c7cb1829de3e9a811536b4ed96b22ce.

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 6]
```diff
@@ -965,20 +965,12 @@ public AndConstraint<TAssertions> ContainItemsAssignableTo<TExpectation>(string
...
-                                "Expected {context:collection} to contain only items of type {0}{reason}" +
-                                ", but item {1} at index {2} is of type {3}.", typeof(TExpectation), item, index, item.GetType());
-                    }
-
-                    ++index;
-                }
+                var actualItems = Subject.ConvertOrCastToCollection();
+                Execute.Assertion
+                    .BecauseOf(because, becauseArgs)
+                    .WithExpectation("Expected {context:collection} to contain element assignable to type {0}{reason}, ", typeof(TExpectation).FullName)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:971]: 🔧 I would rephrase the middle of the message as "to contain at least one element...."

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 7]
```diff
@@ -946,7 +946,7 @@ public AndConstraint<TAssertions> ContainInOrder(params T[] expected)
         }
 
         /// <summary>
-        /// Asserts that the current collection only contains items that are assignable to the type <typeparamref name="TExpectation" />.
+        /// Asserts that the current collection contains items that are assignable to the type <typeparamref name="TExpectation" />.
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:949]: 🔧 I would rephrase the middle of the message as "to contain at least one element...."

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs (hunk 7 above)
Comment (MullerWasHere) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: Done in 3e7255089ab330db576c5c91d5a63daf8f10611a.

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs (hunk 6 above)
Comment (MullerWasHere) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: Done in 3e7255089ab330db576c5c91d5a63daf8f10611a. Refactored method to contain a single assertion to remove _Expectation_'s message duplication.

**Review by dennisdoomen — APPROVED** [review]

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 8]
```diff
@@ -957,29 +957,16 @@ public AndConstraint<TAssertions> ContainInOrder(params T[] expected)
...
-                    }
-
-                    ++index;
-                }
-            }
+                .FailWith("but found <null>.")
+                .Then
+                .Given(() => Subject.ConvertOrCastToCollection())
+                .ForCondition(subject => subject.Any(x => typeof(TExpectation).IsAssignableFrom(GetType(x))))
+                .FailWith("but found {0}.", subject => subject.Select(x => x.GetType()));
```
Comment (jnyrup) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:969]: End method chain with `ClearExpectation()`.

**Review by jnyrup — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs (hunk 8 above)
Comment (MullerWasHere) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: Done in b0afd36b3c1f667250ef0f08dd19fb0107987154.

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 9]
```diff
@@ -957,29 +957,18 @@ public AndConstraint<TAssertions> ContainInOrder(params T[] expected)
...
-                    }
-
-                    ++index;
-                }
-            }
+                .FailWith("but found <null>.")
+                .Then
+                .Given(() => Subject.ConvertOrCastToCollection())
+                .ForCondition(subject => subject.Any(x => typeof(TExpectation).IsAssignableFrom(GetType(x))))
+                .FailWith("but found {0}.", subject => subject.Select(x => x.GetType()))
```
Comment (jnyrup) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:969]: I think you'll need to replace `x.GetType()` with `GetType(x)`. 
Otherwise the failure message of 
```cs
new []{ typeof(int) }.Should().ContainItemsAssignableTo<string>()
```
would contain `RuntimeType` instead of `System.Int32` as we call `typeof(int32).GetType()`.

Comment (MullerWasHere) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: Done in https://github.com/fluentassertions/fluentassertions/pull/1765/commits/65a30b1aa34d735173589b02dc188542ac3f6b51.

**Review by jnyrup — APPROVED** [review]

**Review by dennisdoomen — APPROVED** [review]
